compiler/rustc_ty_utils/src/needs_finalizer.rs

```diff
@@ -0,0 +1,207 @@
+//! Check whether a type has (potentially) non-trivial drop glue.
+
+//! Check whether a type has (potentially) non-trivial drop glue.
+
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::def_id::DefId;
+use rustc_middle::ty::subst::Subst;
+use rustc_middle::ty::util::{needs_drop_components, AlwaysRequiresDrop};
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_session::Limit;
+use rustc_span::{DUMMY_SP};
+
+type NeedsDropResult<T> = Result<T, AlwaysRequiresDrop>;
+
+fn needs_finalizer_raw<'tcx>(tcx: TyCtxt<'tcx>, query: ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool {
+    let finalizer_fields =
+        move |adt_def: &ty::AdtDef| tcx.adt_finalizer_tys(adt_def.did).map(|tys| tys.iter());
+    let res = NeedsDropTypes::new(
+        tcx,
+        query.param_env,
+        query.value,
+        finalizer_fields,
+    )
+    .next()
+    .is_some();
+    debug!("needs_finalize_raw({:?}) = {:?}", query, res);
```

Comment:
Do we want this left in?

Reply:
I know that you're allergic to debug strings, but this stuff is actually really useful when debugging the GC because we can see which types it thought needed finalizing and which didn't without having to keep recompiling rustc.

Reply:
Bah :p